state/applicationofferuser.go

```diff
@@ -61,29 +60,111 @@ func (st *State) UpdateOfferAccess(offer names.ApplicationOfferTag, user names.U
 	if err != nil {
 		return errors.Trace(err)
 	}
-	op := updatePermissionOp(applicationOfferKey(offerUUID), userGlobalKey(userAccessID(user)), access)
 
-	err = st.db().RunTransaction([]txn.Op{op})
-	if err == txn.ErrAborted {
-		return errors.NotFoundf("existing permissions")
+	buildTxn := func(int) ([]txn.Op, error) {
+		_, err := st.GetOfferAccess(offerUUID, user)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		isAdmin, err := st.isControllerOrModelAdmin(user)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ops := []txn.Op{updatePermissionOp(applicationOfferKey(offerUUID), userGlobalKey(userAccessID(user)), access)}
+		if !isAdmin && access != permission.ConsumeAccess && access != permission.AdminAccess {
+			statusOps, err := st.suspendRevokedRelationsOps(offerUUID, user.Id())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ops = append(ops, statusOps...)
+		}
+		return ops, nil
 	}
+
+	err = st.db().Run(buildTxn)
 	return errors.Trace(err)
 }
 
+// suspendRevokedRelationsOps suspends any relations the given user has against
+// the specified offer.
+func (st *State) suspendRevokedRelationsOps(offerUUID, userId string) ([]txn.Op, error) {
+	conns, err := st.OfferConnections(offerUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var ops []txn.Op
+	for _, oc := range conns {
+		if oc.UserName() == userId {
+			rel, err := st.Relation(oc.RelationId())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if rel.Suspended() {
+				continue
+			}
+			suspendOp := txn.Op{
+				C:      relationsC,
+				Id:     rel.doc.DocID,
+				Assert: txn.DocExists,
+				Update: bson.D{{"$set", bson.D{{"suspended", true}}}},
+			}
+			ops = append(ops, suspendOp)
+		}
+	}
+
+	// Add asserts that the relations against the offered application don't change.
+	// This is broader that what we need but's all that's possible.
+	ao := NewApplicationOffers(st)
+	offer, err := ao.ApplicationOfferForUUID(offerUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	app, err := st.Application(offer.ApplicationName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rel, err := app.Relations()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sameRelCount := bson.D{{"relationcount", len(rel)}}
+	ops = append(ops, txn.Op{
```

Comment:
not quite enough. we also need a relationsC txn.Op for each relation we *don't* update, just to make sure they still exist. otherwise one of the old relations could be removed, and a new one added which we would care about

Reply:
done